### PR TITLE
Add placeholder support when targeting Android View host code

### PR DESCRIPTION
**Pull request by veyndan**
A lot of the logic here is just showing the placeholder when the item is loading. However, as the placeholder has to be retrieved asynchronously too, we need a placeholder (`blankView`) for the placeholder.

As the default `RecyclerView` item container has zero-height, `RecyclerView` attempts to load _all_ the available elements, including those that should be off screen (as an infinite number of zero-height items can fit in any `RecyclerView`). To avoid this, we add a `blankView` instead, which is an arbitrarily large, empty `FrameLayout`.

The `blankView` is swapped out with the placeholder once it's ready (i.e., the `onCodeLoaded` callback is invoked). The placeholder is then cached within the `ViewHolder`, so on re-use of the `ViewHolder`, the placeholder can be immediately retrieved and the `blankView` step can be skipped completely.

This will most likely need to be refactored if we want to go down [this route](https://github.com/cashapp/redwood/pull/772#discussion_r1091804731).

### CODE REVIEW

File: redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt [hunk 1]
```diff
@@ -62,64 +54,141 @@ internal class ViewLazyColumn<A : AppService>(
...
+      holder.ziplineScope.close()
+    }
+
+    private class ViewHolder<A : AppService>(
+      val container: FrameLayout,
+      treehouseApp: TreehouseApp<A>,
+      widgetSystem: TreehouseView.WidgetSystem<A>,
+    ) : RecyclerView.ViewHolder(container) {
+
+      val ziplineScope = ZiplineScope()
```
Comment (veyndan) [inline redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt:91]: I'm not adding anything to this yet, because I'm a bit unsure how to do so. Any ideas @swankjesse?

File: redwood-treehouse-host/src/androidMain/kotlin/app/cash/redwood/treehouse/TreehouseWidgetView.kt [hunk 2]
```diff
@@ -35,6 +35,7 @@ import kotlinx.coroutines.flow.StateFlow
 public class TreehouseWidgetView<A : AppService>(
   context: Context,
   override val widgetSystem: TreehouseView.WidgetSystem<A>,
+  override val codeListener: CodeListener = CodeListener(),
```
Comment (swankjesse) [inline redwood-treehouse-host/src/androidMain/kotlin/app/cash/redwood/treehouse/TreehouseWidgetView.kt:38]: Don’t make this change in this PR? It’ll make it more difficult to update

Comment (swankjesse) [reply redwood-treehouse-host/src/androidMain/kotlin/app/cash/redwood/treehouse/TreehouseWidgetView.kt:38]: (Also note that the API works as-is on purpose.)

File: redwood-treehouse-lazylayout-api/src/commonMain/kotlin/app/cash/redwood/treehouse/lazylayout/api/properties.kt [hunk 3]
```diff
@@ -26,14 +26,20 @@ import kotlinx.serialization.modules.contextual
 @Serializable
 public class LazyListIntervalContent(
   @Contextual public val count: Int,
+  @Contextual public val placeholderProvider: Placeholder,
   @Contextual public val itemProvider: Item,
 ) {
 
+  public interface Placeholder : ZiplineService {
```
Comment (swankjesse) [inline redwood-treehouse-lazylayout-api/src/commonMain/kotlin/app/cash/redwood/treehouse/lazylayout/api/properties.kt:33]: Rename to `PlaceholderItem`  ?

File: redwood-treehouse-lazylayout-api/src/commonMain/kotlin/app/cash/redwood/treehouse/lazylayout/api/properties.kt [hunk 4]
```diff
@@ -26,14 +26,20 @@ import kotlinx.serialization.modules.contextual
 @Serializable
 public class LazyListIntervalContent(
   @Contextual public val count: Int,
+  @Contextual public val placeholderProvider: Placeholder,
```
Comment (swankjesse) [inline redwood-treehouse-lazylayout-api/src/commonMain/kotlin/app/cash/redwood/treehouse/lazylayout/api/properties.kt:29]: Rename to placeholderItemProvider?

File: redwood-treehouse-lazylayout-compose/src/jsMain/kotlin/app/cash/redwood/treehouse/lazylayout/compose/TreehouseLazyColumn.kt [hunk 5]
```diff
@@ -36,15 +36,17 @@ public fun ProtocolBridge.LazyColumn(content: LazyListScope.() -> Unit) {
 public interface LazyListScope {
   public fun items(
     count: Int,
+    placeholderContent: @Composable () -> Unit,
```
Comment (swankjesse) [inline redwood-treehouse-lazylayout-compose/src/jsMain/kotlin/app/cash/redwood/treehouse/lazylayout/compose/TreehouseLazyColumn.kt:39]: Rename to placeholderItemContent ?

File: redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt [hunk 6]
```diff
@@ -62,64 +54,141 @@ internal class ViewLazyColumn<A : AppService>(
       layoutParams = ViewGroup.LayoutParams(MATCH_PARENT, WRAP_CONTENT)
     }
     value.adapter = adapter
+    value.recycledViewPool.setMaxRecycledViews(adapter.getItemViewType(RecyclerView.NO_POSITION), 100)
```
Comment (swankjesse) [inline redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt:57]: Revert this change? Or set it to something smaller, maybe 5 or 10 ?

File: redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt [hunk 7]
```diff
@@ -62,64 +54,141 @@ internal class ViewLazyColumn<A : AppService>(
...
+  private class LazyListIntervalContentAdapter<A : AppService>(
     private val treehouseApp: TreehouseApp<A>,
     private val widgetSystem: TreehouseView.WidgetSystem<A>,
-    private val contentHeight: Int,
-  ) : ListAdapter<LazyContentItem, ViewHolder<A>>(LazyContentItemDiffCallback) {
+  ) : ListAdapter<LazyListIntervalContent, LazyListIntervalContentAdapter.ViewHolder<A>>(LazyListIntervalContentDiffCallback) {
     override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder<A> {
       val container = FrameLayout(parent.context).apply {
-        layoutParams = ViewGroup.LayoutParams(MATCH_PARENT, contentHeight)
+        layoutParams = ViewGroup.LayoutParams(MATCH_PARENT, WRAP_CONTENT)
```
Comment (swankjesse) [inline redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt:70]: Nice

File: redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt [hunk 8]
```diff
@@ -62,64 +54,141 @@ internal class ViewLazyColumn<A : AppService>(
...
-    widgetSystem: TreehouseView.WidgetSystem<A>,
-  ) : RecyclerView.ViewHolder(container) {
-    val treehouseWidgetView = TreehouseWidgetView(container.context, widgetSystem)
-      .apply {
-        treehouseApp.renderTo(this)
-
-        layoutParams = FrameLayout.LayoutParams(WRAP_CONTENT, WRAP_CONTENT).apply {
-          gravity = Gravity.CENTER_HORIZONTAL
+    override fun onViewRecycled(holder: ViewHolder<A>) {
+      holder.ziplineScope.close()
```
Comment (swankjesse) [inline redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt:82]: I’m worried that this isn’t safe?

File: redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt (hunk 1 above)
Comment (swankjesse) [reply redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt:91]: Let’s pair on it! Pretty much what we’ll want is to call `item.withZiplineScope(scope).get()`

File: redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt [hunk 9]
```diff
@@ -62,64 +54,141 @@ internal class ViewLazyColumn<A : AppService>(
...
 
   override fun intervals(intervals: List<LazyListIntervalContent>) {
-    adapter.submitList(
-      intervals.flatMap { interval ->
-        List(interval.count) { index ->
-          LazyContentItem(index, interval.itemProvider)
-        }
-      },
-    )
+    adapter.submitList(intervals.flatMap { interval -> List(interval.count) { interval } })
```
Comment (swankjesse) [inline redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt:61]: nit: this is not more readable

File: redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt [hunk 10]
```diff
@@ -62,64 +54,141 @@ internal class ViewLazyColumn<A : AppService>(
...
       return ViewHolder(container, treehouseApp, widgetSystem)
     }
 
     override fun onBindViewHolder(holder: ViewHolder<A>, position: Int) {
-      val itemContent = currentList[position]
-      holder.treehouseWidgetView.setContent {
-        itemContent.item.get(itemContent.index)
-      }
+      holder.itemTreehouseWidgetViewLoadState = LoadState.Loading
+      holder.interval = getItem(position)!!
```
Comment (swankjesse) [inline redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt:77]: Is the variable an _interval_ or is it an _item_ ? Since the position is already applied I don’t think interval makes sense as a name?

File: redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt [hunk 11]
```diff
@@ -62,64 +54,141 @@ internal class ViewLazyColumn<A : AppService>(
...
+    ) : RecyclerView.ViewHolder(container) {
+
+      val ziplineScope = ZiplineScope()
+
+      var interval: LazyListIntervalContent? = null
+
+      val blankView = FrameLayout(itemView.context).apply {
+        layoutParams = FrameLayout.LayoutParams(
+          MATCH_PARENT,
+          TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 1000F, itemView.resources.displayMetrics).toInt(),
```
Comment (swankjesse) [inline redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt:98]: Extract a constant for 1000f ? Like BLANK_ITEM_HEIGHT ?

File: redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt [hunk 12]
```diff
@@ -62,64 +54,141 @@ internal class ViewLazyColumn<A : AppService>(
...
+          placeholderTreehouseWidgetViewLoadState = newState
+          onLoadStateChange()
+        }
+      }
+      val placeholderTreehouseWidgetView: TreehouseWidgetView<A> = TreehouseWidgetView(container.context, widgetSystem, placeholderTreehouseWidgetViewCodeListener)
+        .apply {
+          treehouseApp.renderTo(this)
+
+          layoutParams = FrameLayout.LayoutParams(WRAP_CONTENT, WRAP_CONTENT).apply {
+            gravity = Gravity.CENTER_HORIZONTAL
```
Comment (swankjesse) [inline redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt:116]: What does this do?

File: redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt [hunk 13]
```diff
@@ -62,64 +54,141 @@ internal class ViewLazyColumn<A : AppService>(
...
+          if (placeholderTreehouseWidgetViewLoadState == newState) return
+          placeholderTreehouseWidgetViewLoadState = newState
+          onLoadStateChange()
+        }
+      }
+      val placeholderTreehouseWidgetView: TreehouseWidgetView<A> = TreehouseWidgetView(container.context, widgetSystem, placeholderTreehouseWidgetViewCodeListener)
+        .apply {
+          treehouseApp.renderTo(this)
+
+          layoutParams = FrameLayout.LayoutParams(WRAP_CONTENT, WRAP_CONTENT).apply {
```
Comment (swankjesse) [inline redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt:115]: Should it MATCH_PARENT for the width?

File: redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt [hunk 14]
```diff
@@ -62,64 +54,141 @@ internal class ViewLazyColumn<A : AppService>(
...
+          if (itemTreehouseWidgetViewLoadState == newState) return
+          itemTreehouseWidgetViewLoadState = newState
+          onLoadStateChange()
+        }
       }
+      val itemTreehouseWidgetView: TreehouseWidgetView<A> = TreehouseWidgetView(container.context, widgetSystem, itemTreehouseWidgetViewCodeListener)
+        .apply {
+          treehouseApp.renderTo(this)
+
+          layoutParams = FrameLayout.LayoutParams(WRAP_CONTENT, WRAP_CONTENT).apply {
```
Comment (swankjesse) [inline redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt:133]: MATCH_PARENT for the width?

File: redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt [hunk 15]
```diff
@@ -62,64 +54,141 @@ internal class ViewLazyColumn<A : AppService>(
...
+            gravity = Gravity.CENTER_HORIZONTAL
+          }
+        }
+
+      fun onLoadStateChange() {
+        val interval = checkNotNull(interval)
+
+        val holder: ViewHolder<A> = this
```
Comment (swankjesse) [inline redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt:141]: Inline `holder` everywhere? That one’s my fault

File: redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt [hunk 16]
```diff
@@ -62,64 +54,141 @@ internal class ViewLazyColumn<A : AppService>(
...
+        holder.container.removeAllViews()
+        when {
+          holder.itemTreehouseWidgetViewLoadState == LoadState.Loaded -> {
+            // Show the item
+            holder.itemTreehouseWidgetView.isVisible = true
+            holder.container.addView(holder.itemTreehouseWidgetView)
+          }
+          holder.placeholderTreehouseWidgetViewLoadState == LoadState.Loading -> {
```
Comment (swankjesse) [inline redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt:149]: Reorder the 2nd and 3rd when clauses? That way they’re processed in order of preference. (item, placeholder, blank)

File: redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt [hunk 17]
```diff
@@ -62,64 +54,141 @@ internal class ViewLazyColumn<A : AppService>(
...
+            // Show the blank
+            holder.container.addView(holder.blankView)
+
+            // Start loading the placeholder
```
Comment (swankjesse) [inline redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt:153]: Can we start this load earlier? Like when this `ViewHolder` is created?

File: redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt [hunk 18]
```diff
@@ -62,64 +54,141 @@ internal class ViewLazyColumn<A : AppService>(
...
+            holder.placeholderTreehouseWidgetView.setContent {
+              interval.placeholderProvider.get()
+            }
+          }
+          holder.placeholderTreehouseWidgetViewLoadState == LoadState.Loaded -> {
+            // Show the placeholder.
+            holder.placeholderTreehouseWidgetView.isVisible = true
+            holder.container.addView(holder.placeholderTreehouseWidgetView)
+
+            // Start loading the item
```
Comment (swankjesse) [inline redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt:165]: Can we start this in onBindViewHolder ?

File: samples/emoji-search/presenter/src/commonMain/kotlin/com/example/redwood/emojisearch/presenter/EmojiSearch.kt [hunk 19]
```diff
@@ -102,7 +103,25 @@ fun EmojiSearch(
       hint = "Search",
       onChange = { searchTerm = it },
     )
-    columnProvider.create(filteredEmojis) { image ->
+    columnProvider.create(
+      filteredEmojis,
+      placeholderContent = {
+        val image = EmojiImage(
+          label = "loading…",
+          url = "https://github.githubassets.com/images/icons/emoji/unicode/231a.png?v8",
```
Comment (swankjesse) [inline samples/emoji-search/presenter/src/commonMain/kotlin/com/example/redwood/emojisearch/presenter/EmojiSearch.kt:111]: Add a comment describing which item this is?

**Review by swankjesse — APPROVED** [review]

**Comment by JakeWharton** [comment]
@veyndan safe to close this now?
